synapse/lib/stormlib/vault.py

```diff
@@ -0,0 +1,738 @@
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.cell as s_cell
+import synapse.lib.stormtypes as s_stormtypes
+
+stormcmds = (
+    {
+        'name': 'vault.add',
+        'descr': '''
+            Add a vault.
+
+            Examples:
+
+                // Add a global vault with type `synapse-test`
+                vault.add "shared-global-vault" synapse-test ({'apikey': 'foobar'}) ({}) --scope global
+
+                // Add a user vault with type `synapse-test`
+                vault.add "visi-user-vault" synapse-test ({'apikey': 'barbaz'}) ({}) --scope user --user visi
+
+                // Add a role vault with type `synapse-test`
+                vault.add "contributor-role-vault" synapse-test ({'apikey': 'bazquux'}) ({}) --scope role --user contributor
+
+                // Add an unscoped vault with type `synapse-test`
+                vault.add "unscoped-vault" synapse-test ({'apikey': 'quuxquo'}) ({'server': 'api.foobar.com'})
+        ''',
+        'cmdargs': (
+            ('name', {'type': 'str', 'help': 'The vault name.'}),
+            ('type', {'type': 'str', 'help': 'The vault type.'}),
+            ('secrets', {'help': 'The secrets to store in the new vault.'}),
+            ('configs', {'help': 'The configs to store in the new vault.'}),
+            ('--user', {'type': 'str',
+                        'help': 'This vault is a user-scoped vault, for the specified user name.'}),
+            ('--role', {'type': 'str',
+                        'help': 'This vault is a role-scoped vault, for the specified role name.'}),
+            ('--unscoped', {'type': 'str',
+                            'help': 'This vault is an unscoped vault, for the specified user name.'}),
+            ('--global', {'type': 'bool', 'action': 'store_true', 'default': False,
+                          'help': 'This vault is a global-scoped vault.'}),
+        ),
+        'storm': '''
+            $owner = $lib.null
+            $scope = $lib.null
+
+            $exclusive_args = (0)
+            if $cmdopts.user { $exclusive_args = ($exclusive_args + 1) }
+            if $cmdopts.role { $exclusive_args = ($exclusive_args + 1) }
+            if $cmdopts.unscoped { $exclusive_args = ($exclusive_args + 1) }
+            if $cmdopts.global { $exclusive_args = ($exclusive_args + 1) }
+
+            if ($exclusive_args != 1) {
+                $lib.exit('Must specify one of --user <username>, --role <rolename>, --unscoped <username>, --global.')
+            }
+
+            if $cmdopts.user {
+                $scope = user
+                $owner = $lib.auth.users.byname($cmdopts.user).iden
+            }
+
+            if $cmdopts.role {
+                $scope = role
+                $owner = $lib.auth.roles.byname($cmdopts.role).iden
+            }
+
+            if $cmdopts.unscoped {
+                $owner = $lib.auth.users.byname($cmdopts.user).iden
+            }
+
+            if $cmdopts.global {
```

Comment:
Should this be enforced in the library as well (owner=null if scope=global)?  Or should the owner also be the current user here?

Reply:
I don't think so. The owner/scope checks are done in the cortex API: https://github.com/vertexproject/synapse/blob/blackout/SYN-5887/vaults/synapse/cortex.py#L6736. For global scope vaults, you have to be an admin to create it, then the alluser role gets read access and that's it. Which means only admins can edit global vaults. You can see the permissions being configured for each scope type here: https://github.com/vertexproject/synapse/blob/blackout/SYN-5887/vaults/synapse/cortex.py#L6760